src/parsec.rs

```diff
@@ -247,6 +244,30 @@ impl ParsecMap {
             );
         }
     }
+
+    fn add_new(&mut self, full_id: FullId, gen_pfx_info: &GenesisPfxInfo, log_ident: &LogIdent) {
+        if let Entry::Vacant(entry) = self.map.entry(*gen_pfx_info.first_info.version()) {
+            let _ = entry.insert(create(full_id, gen_pfx_info));
+            self.size_counter = ParsecSizeCounter::default();
+            info!(
+                "{}: Init new Parsec, genesis = {:?}",
+                log_ident, gen_pfx_info
+            );
+        }
+    }
+
+    fn remove_old(&mut self) {
+        // We should only need to remove one parsec at a time, but to make sure we are not leaking
+        // anything allow deleting multiple
+        let mut iter = 0;
+        while self.map.len() > MAX_PARSECS && iter < 100 {
+            if let Some((key, _value)) = self.map.iter_mut().next_back() {
+                let key_copy = *key;
+                let _ = self.map.remove(&key_copy);
+            }
+            iter += 1;
+        }
```

Comment:
Hmm, this loop looks unnecessarily complex to me. Wouldn't something like this work a bit better?

```rust
let parsec_map = mem::replace(&mut self.map, Default::default());
self.map = parsec_map.into_iter().rev().take(MAX_PARSECS).rev().collect();
```

I'm just not sure if `Default::default()` allocates here - I would expect not, but the documentation isn't clear on that.

Reply:
Well it's certainly a more elegant solution, so I'm gonna go with that (your version)!